### ISSUE TITLE
new monotonic broken on docker?

### ISSUE THREAD

**Issue by davidszotten** [eventlet/eventlet#401]
```
$ cat Dockerfile
FROM python:2.7
RUN pip install eventlet==0.21.0
CMD python -c 'import eventlet; eventlet.monkey_patch(); eventlet.listen(("localhost", 8000))'
```

```
$ docker build .
Sending build context to Docker daemon 3.584 kB
Step 1 : FROM python:2.7
 ---> 68caceba17ab
Step 2 : RUN pip install eventlet==0.21.0
 ---> Using cache
 ---> 59c6976d7a1a
Step 3 : RUN python -c 'import eventlet; eventlet.monkey_patch(); eventlet.listen(("localhost", 8000))'
 ---> Running in 1aab15c49d2b
Exception AttributeError: "'_SocketDuckForFd' object has no attribute '_closed'" in <bound method _SocketDuckForFd.__del__ of _SocketDuckForFd:8> ignored
Exception AttributeError: "'_SocketDuckForFd' object has no attribute '_closed'" in <bound method _SocketDuckForFd.__del__ of _SocketDuckForFd:9> ignored
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/eventlet/convenience.py", line 40, in listen
    sock = socket.socket(family, socket.SOCK_STREAM)
  File "/usr/local/lib/python2.7/site-packages/eventlet/greenio/base.py", line 137, in __init__
    notify_opened(fd.fileno())
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 189, in notify_opened
    hub = get_hub()
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 116, in get_hub
    use_hub()
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 70, in use_hub
    mod = get_default_hub()
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 38, in get_default_hub
    import eventlet.hubs.epolls
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/epolls.py", line 27, in <module>
    from eventlet.hubs.hub import BaseHub
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 23, in <module>
    from eventlet.support import greenlets as greenlet, clear_sys_exc_info, monotonic, six
  File "/usr/local/lib/python2.7/site-packages/eventlet/support/monotonic.py", line 167, in <module>
    raise RuntimeError('no suitable implementation for this system')
RuntimeError: no suitable implementation for this system
```

**Comment by mvdbertram** [comment]
same error happens also when running not within docker, but also on debian jessie

**Comment by temoto** [comment]
Thanks for reproduction script.

**Comment by temoto** [comment]
The real issue, of course, has nothing to do with monotonic. But it does false help by masking real exception. Real issue is import cycle.

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/eventlet/convenience.py", line 40, in listen
    sock = socket.socket(family, socket.SOCK_STREAM)
  File "/usr/local/lib/python2.7/site-packages/eventlet/greenio/base.py", line 137, in __init__
    notify_opened(fd.fileno())
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 189, in notify_opened
    hub = get_hub()
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 116, in get_hub
    use_hub()
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 70, in use_hub
    mod = get_default_hub()
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 38, in get_default_hub
    import eventlet.hubs.epolls
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/epolls.py", line 27, in <module>
    from eventlet.hubs.hub import BaseHub
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 23, in <module>
    from eventlet.support import greenlets as greenlet, clear_sys_exc_info, monotonic, six
  File "/usr/local/lib/python2.7/site-packages/eventlet/support/monotonic.py", line 137, in <module>
    clock_gettime = ctypes.CDLL(ctypes.util.find_library('rt'),
  File "/usr/local/lib/python2.7/ctypes/util.py", line 274, in find_library
    return _findSoname_ldconfig(name) or _get_soname(_findLib_gcc(name))
  File "/usr/local/lib/python2.7/ctypes/util.py", line 264, in _findSoname_ldconfig
    env=env)
  File "/usr/local/lib/python2.7/site-packages/eventlet/green/subprocess.py", line 55, in __init__
    subprocess_orig.Popen.__init__(self, args, 0, *argss, **kwds)
  File "/usr/local/lib/python2.7/subprocess.py", line 420, in __init__
    self.stdout = os.fdopen(c2pread, 'rb', bufsize)
  File "/usr/local/lib/python2.7/site-packages/eventlet/green/os.py", line 28, in fdopen
    return greenio.GreenPipe(fd, *args, **kw)
  File "/usr/local/lib/python2.7/site-packages/eventlet/greenio/py2.py", line 41, in __init__
    super(GreenPipe, self).__init__(_SocketDuckForFd(fileno), mode, bufsize)
  File "/usr/local/lib/python2.7/site-packages/eventlet/greenio/py2.py", line 128, in __init__
    notify_opened(fileno)
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 189, in notify_opened
    hub = get_hub()
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 116, in get_hub
    use_hub()
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 70, in use_hub
    mod = get_default_hub()
  File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 39, in get_default_hub
    return eventlet.hubs.epolls
AttributeError: 'module' object has no attribute 'epolls'
```

**Comment by temoto** [comment]
Meanwhile, dirty workaround is `eventlet.sleep()` before `monkey_patch()`.

```diff
import eventlet
+eventlet.sleep()
eventlet.monkey_patch()
eventlet.listen(("localhost", 8000))
```

**Comment by pttp** [comment]
can this workaround be used for gunicorn eventlet class workers?
how would you fix those?

**Comment by pttp** [comment]
although that's probably not really useful, I can confirm that this workaround also would fix gunicorn eventlet workers..

adding `    eventlet.sleep()` directly after `    import eventlet` in line 11 in https://github.com/benoitc/gunicorn/blob/master/gunicorn/workers/geventlet.py can integrate this workaround to the gunicorn worker spawn process.. 

anyway, that's not really applicable... this should be fixed in eventlet

**Comment by temoto** [comment]
Yeah should be fixed here.

**Comment by rburkholder** [comment]
Just in case a further reference to this issue is needed, here is another re-creation, this using Ryu 4.13, and a code evaluation.  Will the eventless.sleep() change work here?  In which file is that change inserted?

Error message: AttributeError("'module' object has no attribute 'epolls'",)

Possible explanation:

[urllib3 is intercepting  select calls](https://github.com/shazow/urllib3/issues/1104), and not providing epoll or epolls.

Question:

How to bypass this?  My script worked a few months ago when testing, so I am trying to narrow down what changed.

In code in monotonic.py, I changed to get some debugging output (exception occurs on the statement between 'line d' and 'line g'), following code is starting around line 132 in montonic.py:
```
        else:
            print "line b"
            print sys.path
            print ctypes
            print ctypes.util
            try:
                print "line d"
                print ctypes.util.find_library('c')
                print "line g"
                clock_gettime = ctypes.CDLL(ctypes.util.find_library('c'),
                                            use_errno=True).clock_gettime
            except AttributeError as e:
                print( "line e ", e )
                clock_gettime = ctypes.CDLL(ctypes.util.find_library('rt'),
                                            use_errno=True).clock_gettime
                print "line f"
```
search for 'line e' in the following output to spot the module with problems.


>root@edge:/home/vagrant# ryu run --observe-links  /usr/local/lib/python2.7/dist-packages/ryu/app/gui_topology/gui_topology.py --install-lldp-flow 
loading app /vagrant/test1.py 
loading app ryu.topology.switches 
loading app ryu.controller.dpset 
loading app ryu.controller.ofp_handler 
instantiating app ryu.controller.dpset of DPSet 
instantiating app ryu.topology.switches of Switches 
line a 
line b 
['/usr/share/python-wheels/CacheControl-0.11.7-py2.py3-none-any.whl', '/usr/share/python-wheels/appdirs-1.4.0-py2.py3-none-any.whl', '/usr/share/python-wheels/chardet-2.3.0-py2.py3-none-any.whl', '/usr/share/python-wheels/colorama-0.3.7-py2.py3-none-any.whl', '/usr/share/python-wheels/distlib-0.2.4-py2.py3-none-any.whl', '/usr/share/python-wheels/distro-1.0.1-py2.py3-none-any.whl', '/usr/share/python-wheels/html5lib-0.999999999-py2.py3-none-any.whl', '/usr/share/python-wheels/ipaddress-0.0.0-py2.py3-none-any.whl', '/usr/share/python-wheels/lockfile-0.12.2-py2.py3-none-any.whl', '/usr/share/python-wheels/packaging-16.8-py2.py3-none-any.whl', '/usr/share/python-wheels/pip-9.0.1-py2.py3-none-any.whl', '/usr/share/python-wheels/pkg_resources-0.0.0-py2.py3-none-any.whl', '/usr/share/python-wheels/progress-1.2-py2.py3-none-any.whl', '/usr/share/python-wheels/pyparsing-2.1.10-py2.py3-none-any.whl', '/usr/share/python-wheels/requests-2.12.4-py2.py3-none-any.whl', '/usr/share/python-wheels/retrying-1.3.3-py2.py3-none-any.whl', '/usr/share/python-wheels/setuptools-32.3.1-py2.py3-none-any.whl', '/usr/share/python-wheels/six-1.10.0-py2.py3-none-any.whl', '/usr/share/python-wheels/urllib3-1.19.1-py2.py3-none-any.whl', '/usr/share/python-wheels/webencodings-0.5-py2.py3-none-any.whl', '/usr/share/python-wheels/wheel-0.29.0-py2.py3-none-any.whl', '/usr/local/bin', '/usr/lib/python2.7', '/usr/lib/python2.7/plat-x86_64-linux-gnu', '/usr/lib/python2.7/lib-tk', '/usr/lib/python2.7/lib-old', '/usr/lib/python2.7/lib-dynload', '/usr/local/lib/python2.7/dist-packages', '/usr/lib/python2.7/dist-packages']
<module 'ctypes' from '/usr/lib/python2.7/ctypes/__init__.pyc'>
<module 'ctypes.util' from '/usr/lib/python2.7/ctypes/util.pyc'>
line d
('line e ', AttributeError("'module' object has no attribute 'epolls'",)) 
Exception AttributeError: "'_SocketDuckForFd' object has no attribute '_closed'" in <bound method _SocketDuckForFd.__del__ of _SocketDuckForFd:7> ignored 
Exception AttributeError: "'_SocketDuckForFd' object has no attribute '_closed'" in <bound method _SocketDuckForFd.__del__ of _SocketDuckForFd:8> ignored Traceback (most recent call last):

>  File "/usr/local/bin/ryu", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/dist-packages/ryu/cmd/ryu_base.py", line 74, in main
    subcmd.run(base_conf.subcommand_args)
  File "/usr/local/lib/python2.7/dist-packages/ryu/cmd/ryu_base.py", line 57, in run
    self.entry(args=args, prog=prog)
  File "/usr/local/lib/python2.7/dist-packages/ryu/cmd/manager.py", line 81, in main
    services.extend(app_mgr.instantiate_apps(**contexts))
  File "/usr/local/lib/python2.7/dist-packages/ryu/base/app_manager.py", line 509, in instantiate_apps
    self._instantiate(app_name, cls, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/ryu/base/app_manager.py", line 495, in _instantiate
    app = cls(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/ryu/topology/switches.py", line 532, in __init__
    self.threads.append(hub.spawn(self.lldp_loop))
  File "/usr/local/lib/python2.7/dist-packages/ryu/lib/hub.py", line 68, in spawn
    return eventlet.spawn(_launch, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/eventlet/greenthread.py", line 50, in spawn
    hub = hubs.get_hub()
  File "/usr/local/lib/python2.7/dist-packages/eventlet/hubs/__init__.py", line 116, in get_hub
    use_hub()
  File "/usr/local/lib/python2.7/dist-packages/eventlet/hubs/__init__.py", line 70, in use_hub
    mod = get_default_hub()
  File "/usr/local/lib/python2.7/dist-packages/eventlet/hubs/__init__.py", line 38, in get_default_hub
    import eventlet.hubs.epolls
  File "/usr/local/lib/python2.7/dist-packages/eventlet/hubs/epolls.py", line 27, in <module>
    from eventlet.hubs.hub import BaseHub
  File "/usr/local/lib/python2.7/dist-packages/eventlet/hubs/hub.py", line 23, in <module>
    from eventlet.support import greenlets as greenlet, clear_sys_exc_info, monotonic, six
  File "/usr/local/lib/python2.7/dist-packages/eventlet/support/monotonic.py", line 178, in <module>
    raise RuntimeError('no suitable implementation for this system')
RuntimeError: no suitable implementation for this system

**Comment by temoto** [comment]
@rburkholder 

> Will the eventless.sleep() change work here?

Try and see.

> In which file is that change inserted?

Where `import eventlet` is executed first time.

**Comment by rburkholder** [comment]
Making the following change to /usr/local/lib/python2.7/dist-packages/ryu/lib/hub.py eliminated the exception (add 'eventless.sleep()) to line 32 of the file after the import statement):

```
if HUB_TYPE == 'eventlet':
    import eventlet
    eventlet.sleep()
    import eventlet.event
    import eventlet.queue
    import eventlet.semaphore
    import eventlet.timeout
    import eventlet.wsgi
```

**Comment by dkavanagh** [comment]
Ran into this as well (via gunicorn on CentOS 7). Verified patch in geventlet.py, but would prefer eventlet release to fix this.

**Comment by enricobacis** [comment]
I confirm this also happens on kali linux (debian-testing based)

**Comment by j-ramirez-bi4group** [comment]
Any news?

**Comment by temoto** [comment]
Not yet.

**Comment by Barlog951** [comment]
This happens me in my docker instance of python 2.7.13 with nameko package

PLS fix this because ist very important for me 

thanks

**Comment by LukeusMaximus** [comment]
I also get this inside of a Docker instance, specifically when running RYU.

**Comment by prometheanfire** [comment]
Are there any updates here or just workarounds?  We are looping to block the new eventlet in openstack due to this bug.  https://review.openstack.org/464687

**Comment by temoto** [comment]
There's a fix but we formed an evil coalition to hide it. Join or perish.

**Comment by marco-lavagnino** [comment]
if you're using libraries and you can't call  eventlet.sleep() just install eventlet==0.20, worked for me!

**Comment by ToastyBiggums** [comment]
This bug has been open this long without a fix? Eventlet developers sleeping?

**Comment by jstasiak** [comment]
> This bug has been open this long without a fix? Eventlet developers sleeping?

@ToastyBiggums To the best of my knowledge Eventlet is a volunteer-based project right now and there's no money changing hands.

I believe it should be clear that unless a suitable patch is provided there's no expectation that any bug is fixed. Some bugs do get fixed but only thanks to people who spend their otherwise free time on it, I certainly wouldn't expect or demand anything of them.

**Comment by zancas** [comment]
Hi folks, I have the same traceback.
I can reproduce this error with:

eventlet==0.21.0, and python==2.7.13

and have verified that this error *DOES NOT* occur against:

eventlet==0.21.0, and python==2.7.12

Soo...   the root cause is a change between python 2.7.12 and 2.7.13.

**Comment by zancas** [comment]
A way forward (with respect to debugging) would be a `git bisect` against the commits from 2.7.12 and 2.7.13.

**Comment by temoto** [comment]
@zancas thank you very much. I was wondering how it stopped reproducing on my test machine.

**Comment by richmondwang** [comment]
I have this issue and I have 
```
eventlet==0.21.0
python==2.7.12
```

But all is well when im using `python==2.7.10`.

**Comment by antocuni** [comment]
As other said, the real issue is that `eventlet==0.21` is broken on `CPython>=2.7.13`. More details here:
https://github.com/eventlet/eventlet/issues/428

**Comment by geofft** [comment]
Possible fix in #434: call `get_hub()` before monkey-patching anything. I've tested that this fixes the minimal reproducer @antocuni posted in #428. You can also implement this workaround in your own code with

```diff
 import eventlet
+eventlet.hubs.get_hub()
 eventlet.monkey_patch(...)
```

which also gets my code working again.

FWIW, the root cause is that in Python 2.7.13, `ctypes.util.find_library` is now using `subprocess.Popen`, instead of `os.popen` (python/cpython@eb063011), and eventlet monkey-patches `subprocess.Popen` but not `os.popen`. So if importing the hub calls `ctypes.util.find_library`, that will trigger a recursive import of the hub, which fails. In particular, the `monotonic` module uses ctypes to try to find monotonic clocks in the platform libc, but this problem isn't specific to `monotonic` at all.

**Comment by temoto** [comment]
@geofft thank you very much. I'm trying to avoid import inside `monkey_patch` function before release.

To everyone: instructions for installing `get_hub` workaround:

`pip install https://github.com/eventlet/eventlet/archive/e2884ef9265024d6580eefdd6aa4ec009743a2d7.zip`

**Comment by geofft** [comment]
You can structure this to avoid an `import` call - just move `import eventlet` to the top of `patcher.py`, and call it by its full name inside `monkey_patch`, `eventlet.hubs.get_hub()`.

**Comment by temoto** [comment]
Right, thank you twice!

**Comment by temoto** [comment]
Everyone: Geoff's patch was merged in master, please tell if it works for you.

`pip install https://github.com/eventlet/eventlet/archive/b756447bab51046dfc6f1e0e299cc997ab343701.zip`

**Comment by jwg4** [comment]
I tested this with the version linked in @temoto's comment - booting eventlet workers for gunicorn still fails for me.

```
(venv)jack@neptune-vm:~/git/ss_data_api/api$ python --version
Python 2.7.12
(venv)jack@neptune-vm:~/git/ss_data_api/api$ pip freeze
api-tools==1.3.1.1
certifi==2017.7.27.1
chardet==3.0.4
click==6.7
enum-compat==0.0.2
enum34==1.1.6
eventlet==0.21.0
Flask==0.12.2
Flask-Cache==0.13.1
Flask-SQLAlchemy==2.2
gevent==1.1.1
greenlet==0.4.12
gunicorn==19.7.1
idna==2.6
itsdangerous==0.24
Jinja2==2.9.6
locustio==0.7.5
MarkupSafe==1.0
msgpack-python==0.4.8
numpy==1.13.1
pandas==0.20.3
prm-dates==0.5.1
psycopg2==2.7.3
python-dateutil==2.6.1
pytz==2017.2
requests==2.18.4
six==1.10.0
SQLAlchemy==1.1.13
urllib3==1.22
Werkzeug==0.12.2
```

**Comment by temoto** [comment]
@jwg4 but AFAIK Python 2.7.12 was unaffected by this problem in first place. Could you post reproduction history from clean virtualenv?

**Comment by jwg4** [comment]
Will do. Perhaps the gunicorn problem is related but slightly different?

On 5 Sep 2017 6:48 pm, "Sergey Shepelev" <notifications@github.com> wrote:

@jwg4 <https://github.com/jwg4> but AFAIK Python 2.7.12 was unaffected by
this problem in first place. Could you post reproduction history from clean
virtualenv?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/eventlet/eventlet/issues/401#issuecomment-327267809>,
or mute the thread

**Comment by jwg4** [comment]
I can confirm that gunicorn with eventlet workers fails with both stock eventlet 0.21.0 and with https://github.com/eventlet/eventlet/archive/b756447bab51046dfc6f1e0e299cc997ab343701.zip as suggested above.

```
(venv) jack@neptune-vm:~/z/eventlet$ pip freeze
enum-compat==0.0.2
enum34==1.1.6
eventlet==0.21.0
greenlet==0.4.12
gunicorn==19.7.1
(venv) jack@neptune-vm:~/z/eventlet$ python --version
Python 2.7.12
```

The command which failed was `gunicorn -w 4 myapp:app -k eventlet`. Without eventlet workers, eg `gunicorn -w 4 myapp:app` succeeded.

Error messages:
```
[2017-09-06 10:42:49 +0000] [18816] [INFO] Starting gunicorn 19.7.1
[2017-09-06 10:42:49 +0000] [18816] [INFO] Listening at: http://127.0.0.1:8000 (18816)
[2017-09-06 10:42:49 +0000] [18816] [INFO] Using worker: eventlet
[2017-09-06 10:42:49 +0000] [18824] [INFO] Booting worker with pid: 18824
[2017-09-06 10:42:49 +0000] [18824] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/gunicorn/arbiter.py", line 578, in spawn_worker
    worker.init_process()
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/gunicorn/workers/geventlet.py", line 102, in init_process
    self.patch()
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/gunicorn/workers/geventlet.py", line 91, in patch
    hubs.use_hub()
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 70, in use_hub
    mod = get_default_hub()
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 38, in get_default_hub
    import eventlet.hubs.epolls
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/eventlet/hubs/epolls.py", line 27, in <module>
    from eventlet.hubs.hub import BaseHub
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 23, in <module>
    from eventlet.support import greenlets as greenlet, clear_sys_exc_info, monotonic, six
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/eventlet/support/monotonic.py", line 169, in <module>
    raise RuntimeError('no suitable implementation for this system')
RuntimeError: no suitable implementation for this system
[2017-09-06 10:42:49 +0000] [18824] [INFO] Worker exiting (pid: 18824)
[2017-09-06 10:42:49 +0000] [18827] [INFO] Booting worker with pid: 18827
[2017-09-06 10:42:49 +0000] [18827] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/gunicorn/arbiter.py", line 578, in spawn_worker
    worker.init_process()
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/gunicorn/workers/geventlet.py", line 102, in init_process
    self.patch()
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/gunicorn/workers/geventlet.py", line 91, in patch
    hubs.use_hub()
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 70, in use_hub
    mod = get_default_hub()
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 38, in get_default_hub
    import eventlet.hubs.epolls
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/eventlet/hubs/epolls.py", line 27, in <module>
    from eventlet.hubs.hub import BaseHub
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 23, in <module>
    from eventlet.support import greenlets as greenlet, clear_sys_exc_info, monotonic, six
  File "/home/jack/z/eventlet/venv/local/lib/python2.7/site-packages/eventlet/support/monotonic.py", line 169, in <module>
    raise RuntimeError('no suitable implementation for this system')
RuntimeError: no suitable implementation for this system
[2017-09-06 10:42:49 +0000] [18827] [INFO] Worker exiting (pid: 18827)
[2017-09-06 10:42:49 +0000] [18816] [INFO] Shutting down: Master
[2017-09-06 10:42:49 +0000] [18816] [INFO] Reason: Worker failed to boot.
```

**Comment by jwg4** [comment]
The entry point to eventlet seems to be `hubs.use_hub()`, rather than `monkey_patch, which might trigger a different but related circular import, one which wasn't fixed by the patch #434

**Comment by geofft** [comment]
Yes, this problem is unrelated and I think it is gunicorn's fault, not eventlet's.

I can reproduce the above with Python 2.7.9 (Debian Jessie, sample app from the gunicorn.org home page). If you remove monotonic's `except Exception`, the actual exception is

```
Traceback (most recent call last):
  File "/tmp/401/local/lib/python2.7/site-packages/gunicorn/arbiter.py", line 578, in spawn_worker
    worker.init_process()
  File "/tmp/401/local/lib/python2.7/site-packages/gunicorn/workers/geventlet.py", line 102, in init_process
    self.patch()
  File "/tmp/401/local/lib/python2.7/site-packages/gunicorn/workers/geventlet.py", line 91, in patch
    hubs.use_hub()
  File "/tmp/401/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 70, in use_hub
    mod = get_default_hub()
  File "/tmp/401/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 38, in get_default_hub
    import eventlet.hubs.epolls
  File "/tmp/401/local/lib/python2.7/site-packages/eventlet/hubs/epolls.py", line 27, in <module>
    from eventlet.hubs.hub import BaseHub
  File "/tmp/401/local/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 23, in <module>
    from eventlet.support import greenlets as greenlet, clear_sys_exc_info, monotonic, six
  File "/tmp/401/local/lib/python2.7/site-packages/eventlet/support/monotonic.py", line 134, in <module>
    clock_gettime = ctypes.CDLL(ctypes.util.find_library('c'),
  File "/usr/lib/python2.7/ctypes/util.py", line 253, in find_library
    return _findSoname_ldconfig(name) or _get_soname(_findLib_gcc(name))
  File "/usr/lib/python2.7/ctypes/util.py", line 246, in _findSoname_ldconfig
    f.close()
IOError: [Errno 10] No child processes
```

That part of ctypes/util.py is calling `os.popen` to run an external process and get the results. If you use strace, it looks like someone is calling `wait4(-1, ...)` (wait for all child processes) before `os.popen` gets to calling `wait4(pid, ...)` on the specific pid. Since the child process has already been waited for, the wait system call returns `ECHILD`.

If you add [a wrapper around the gunicorn process to send SIGSYS if anything tries to `wait(-1)`](https://gist.github.com/geofft/257a9a8b09474d6a6490f7ece4761126) and modify bin/gunicorn to catch SIGSYS and raise a Python exception, the stack trace gains these lines:

```
  File "/usr/lib/python2.7/ctypes/util.py", line 244, in _findSoname_ldconfig
    data = f.read()
  File "/tmp/401/local/lib/python2.7/site-packages/gunicorn/arbiter.py", line 244, in handle_chld
    self.reap_workers()
  File "/tmp/401/local/lib/python2.7/site-packages/gunicorn/arbiter.py", line 512, in reap_workers
    wpid, status = os.waitpid(-1, os.WNOHANG)
  File "bin/gunicorn", line 8, in handler
    raise RuntimeError("signal caught")
```

which points to this code in gunicorn/arbiter.py:

```python
    def handle_chld(self, sig, frame):
        "SIGCHLD handling"
        self.reap_workers()
        self.wakeup()
```

That signal handler probably should not be active when imports are running (which includes when `eventlet.hubs.use_hub()` is being called), or `self.reap_workers()` should be more clever.

Someone who knows more about gunicorn than me should probably raise this as a bug on the gunicorn bug tracker. I don't think there's anything for eventlet to do here.

**Comment by jwg4** [comment]
Thanks! I will try and see if I can isolate this as a gunicorn bug and report it.

**Comment by jwg4** [comment]
No worries, I think I have it: gunicorn waits on all child processes, in order to do something legitimate, probably related to managing multiple worker processes. However, it shouldn't do this when calling out to third-party code which might do things like blocking IO or forking another process, because it will mess up that code's signal handling.

Before the monotonic changes, eventlet typically didn't do anything which triggered this failure (although gunicorn was vulnerable that any worker init code could have done something like this), but now it does.

Is this right?

**Comment by geofft** [comment]
The change between eventlet 0.20 and 0.21 is that it uses (and vendors) monotonic; see #303 and #388. On GNU/Linux, `import monotonic` calls `ctypes.util.find_library`, which in turn shells out to `ldconfig -p` to list the libraries on the system

(On Python < 2.7.13, `ctypes.util.find_library` uses `os.popen`; on 2.7.13 it uses `subprocess.Popen`. That difference was the one that caused the original breakage reported in this ticket / #428 and fixed in #434, but your problem arises with either `os.popen` or `subprocess.Popen`.)

And yes, the problem is that gunicorn waits on all child processes to manage workers, but that setup, including the handler for SIGCHLD, is already in place when it calls `eventlet.hubs.use_hub()`, which triggers an import of monotonic. I don't know if there's a rule that gunicorn workers can't create child processes. If so, it would make sense that forking a child process doesn't work right, but that should only start to apply once the workers actually start, not when gunicorn is setting up eventlet.

One workaround that could be implemented in eventlet alone is to `import monotonic` when eventlet is being imported, instead of waiting until the default hub is loaded. I'm not sure if that is a cleaner solution. (Arguably the real problem is that `import monotonic` goes and spawns a child process, maybe that can just be fixed.)

**Comment by temoto** [comment]
@geofft thank you for detailed explanation.

I agree that spawning process in `find_library` is ugly. Hopefully ctypes developers considered it and made best decision. But it's only ugly, everything should work with a child process here and there. I think gunicorn `wait(-1)` for processes it didn't spawn is the real offender here.

That said, if early import of monotonic could help - that'd be good for everyone and cheap.

@jwg4 please, try again with this patch

`pip install https://github.com/eventlet/eventlet/archive/82f1877ff5b950e1ee9debbb8be880ebedb8abcb.zip`

**Comment by jwg4** [comment]
@temoto It seems to work correctly with the patch, thanks. As you know I also submitted a patch to gunicorn at https://github.com/benoitc/gunicorn/pull/1585 which apparently solves the issue.

**Comment by jwg4** [comment]
It is clear that the gunicorn issue is a different one which leads to the same error message only because they both happen within the same try except. Should I create a separate issue for this so that any further discussion can be separate? Or is this unnecessary?

**Comment by temoto** [comment]
@jwg4 done

**Comment by prometheanfire** [comment]
just to be clear, was https://github.com/eventlet/eventlet/commit/b756447bab51046dfc6f1e0e299cc997ab343701 the fix for this bug?

**Comment by temoto** [comment]
Yes, workaround.

Proper fix requires large rearrangement of modules to allow better import
graph.

**Comment by prometheanfire** [comment]
cool, thanks, do you know when the next release is supposed to be?  I ask because I'd like to uncap eventlet in requirements for openstack.

https://github.com/openstack/requirements/blob/master/global-requirements.txt#L56

**Comment by temoto** [comment]
@prometheanfire I want to fix this https://github.com/eventlet/eventlet/issues/147 problem with multiprocessing, seems it's a huge pain for many people.

Meanwhile, my constant recommendation is to require particular commit id. It has all benefits of specifying exact version (repeatable build, no new errors) + absence of stupid pypi release bureaucracy. The only better way is to include dependency sources within your project to protect against upstream servers downtime.

**Comment by prometheanfire** [comment]
Unfortunately we cannot lock to a sha with how we are organized now,  we only include releases, even rc and beta releases are rejected.

**Comment by temoto** [comment]
@prometheanfire this fix was included in 0.22 release on pypi

**Comment by prometheanfire** [comment]
thanks :D

**Comment by toabctl** [comment]
I still see this problem with eventlet==0.24.1 and monotonic==1.5 :+1: 

```
[   58s] Captured daemon_log:
[   58s] ~~~~~~~~~~~~~~~~~~~~
[   58s]     Exception AttributeError: "'_SocketDuckForFd' object has no attribute '_closed'" in <bound method _SocketDuckForFd.__del__ of _SocketDuckForFd:16> ignored
[   58s]     Exception AttributeError: "'_SocketDuckForFd' object has no attribute '_closed'" in <bound method _SocketDuckForFd.__del__ of _SocketDuckForFd:17> ignored
[   58s]     Traceback (most recent call last):
[   58s]       File "oslo_rootwrap/tests/run_daemon.py", line 20, in <module>
[   58s]         from oslo_rootwrap import cmd
[   58s]       File "/home/abuild/rpmbuild/BUILD/oslo.rootwrap-5.15.2/oslo_rootwrap/__init__.py", line 19, in <module>
[   58s]         import eventlet.patcher
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/__init__.py", line 10, in <module>
[   58s]         from eventlet import convenience
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/convenience.py", line 7, in <module>
[   58s]         from eventlet.green import socket
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/green/socket.py", line 21, in <module>
[   58s]         from eventlet.support import greendns
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/support/greendns.py", line 64, in <module>
[   58s]         setattr(dns, pkg, import_patched('dns.' + pkg))
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/support/greendns.py", line 59, in import_patched
[   58s]         return patcher.import_patched(module_name, **modules)
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/patcher.py", line 120, in import_patched
[   58s]         *additional_modules + tuple(kw_additional_modules.items()))
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/patcher.py", line 94, in inject
[   58s]         module = __import__(module_name, {}, {}, module_name.split('.')[:-1])
[   58s]       File "/usr/lib/python2.7/site-packages/dns/dnssec.py", line 482, in <module>
[   58s]         from Crypto.Hash import MD5, SHA1, SHA256, SHA384, SHA512
[   58s]       File "/usr/lib64/python2.7/site-packages/Crypto/Hash/MD5.py", line 47, in <module>
[   58s]         """)
[   58s]       File "/usr/lib64/python2.7/site-packages/Crypto/Util/_raw_api.py", line 296, in load_pycryptodome_raw_lib
[   58s]         cdecl)
[   58s]       File "/usr/lib64/python2.7/site-packages/Crypto/Util/_raw_api.py", line 175, in load_lib
[   58s]         bits, linkage = platform.architecture()
[   58s]       File "/usr/lib64/python2.7/platform.py", line 1066, in architecture
[   58s]         output = _syscmd_file(executable, '')
[   58s]       File "/usr/lib64/python2.7/platform.py", line 1009, in _syscmd_file
[   58s]         stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
[   58s]       File "/usr/lib64/python2.7/subprocess.py", line 420, in __init__
[   58s]         self.stdout = os.fdopen(c2pread, 'rb', bufsize)
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/green/os.py", line 28, in fdopen
[   58s]         return greenio.GreenPipe(fd, *args, **kw)
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/greenio/py2.py", line 42, in __init__
[   58s]         super(GreenPipe, self).__init__(_SocketDuckForFd(fileno), mode, bufsize)
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/greenio/py2.py", line 129, in __init__
[   58s]         notify_opened(fileno)
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 190, in notify_opened
[   58s]         hub = get_hub()
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 117, in get_hub
[   58s]         use_hub()
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 71, in use_hub
[   58s]         mod = get_default_hub()
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 39, in get_default_hub
[   58s]         import eventlet.hubs.epolls
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/hubs/epolls.py", line 13, in <module>
[   58s]         from eventlet.hubs.hub import BaseHub
[   58s]       File "/usr/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 24, in <module>
[   58s]         import monotonic
[   58s]       File "/usr/lib/python2.7/site-packages/monotonic.py", line 169, in <module>
[   58s]         raise RuntimeError('no suitable implementation for this system: ' + repr(e))
[   58s]     RuntimeError: no suitable implementation for this system: AttributeError("'module' object has no attribute 'epolls'",)
[   58s]     Exception AttributeError: "'_SocketDuckForFd' object has no attribute '_closed'" in <bound method _SocketDuckForFd.__del__ of _SocketDuckForFd:10> ignored
[   58s]     
```

**Comment by dtantsur** [comment]
Still reproducible with eventlet==0.24.1 and monotonic==1.5: https://541a97755748b4fa92b0-120a9aa299e111f1c1833809030ff629.ssl.cf2.rackcdn.com/703258/1/check/bifrost-integration-tinyipa-ubuntu-xenial/5912fee/logs/ironic-inspector.log

**Comment by temoto** [comment]
@toabctl @dtantsur this crutch should work:

Execute this line before importing eventlet.
```diff
+import dns  # workaround for https://github.com/eventlet/eventlet/issues/401
 import eventlet
```

Root cause is import cycle: eventlet -> dns -> Crypto.Hash -> platform -> [patched]subprocess -> eventlet

**Comment by dtantsur** [comment]
> Root cause is import cycle: eventlet -> dns -> Crypto.Hash -> platform -> [patched]subprocess -> eventlet

@temoto thanks for looking into this! Which versions of these packages do you use to confirm the bug? I have two CI jobs that differ in the installed packages, but these ones seem to be the same in both. One works, one crashes with the traceback above..

**Comment by dtantsur** [comment]
I have a suspect: cryptography 2.6.1 on the working run vs 2.8 on the broken one.

**Comment by temoto** [comment]
@dtantsur I didn't run code this time, just read posted logs.

Probably you are looking for packages providing `dns`, `Crypto` or `Cryptodome`.

**Comment by temoto** [comment]
With big no-blame note, you could say this break was provoked **not caused** by this change https://github.com/Legrandin/pycryptodome/commit/f08e5f950656314e31a1f2173e191458d09c0578

**Comment by dtantsur** [comment]
Sounds entirely plausible.

**Comment by chenxing0407** [comment]
you can add traceback to see the real reason, I encounter 
"IOError: [Errno 13] Permission denied: '/dev/null'"
